packages/ember-toucan-core/src/components/form/controls/combobox.ts

```diff
@@ -0,0 +1,537 @@
+   */
+  get activeDescendant() {
+    // For the case where nothing is selected on render
+    if (this.activeIndex === null) {
+      return null;
+    }
+
+    return `${this.popoverId}-${this.activeIndex}`;
+  }
+
+  /**
+   * This state is used to determine if we should add event handlers to the input element or not.
+   */
+  get isDisabledOrReadOnlyOrWithoutOptions() {
+    return (
+      this.args.isDisabled ||
+      this.args.isReadOnly ||
+      this.args.options === undefined
+    );
+  }
+
+  /**
+   * We apply different styles to the input tag based on our current state.
+   */
+  get styles() {
+    if (this.args.isDisabled) {
+      return 'shadow-focusable-outline bg-overlay-1 text-disabled pointer-events-none placeholder:text-disabled';
+    }
+
+    if (this.args.isReadOnly) {
+      return 'focus:shadow-focus-outline bg-surface-xl shadow-read-only-outline text-titles-and-attributes placeholder:text-titles-and-attributes';
+    }
+
+    if (this.args.hasError) {
+      return 'shadow-error-outline focus:shadow-error-focus-outline bg-overlay-1 text-titles-and-attributes';
+    }
+
+    return 'shadow-focusable-outline focus:shadow-focus-outline bg-overlay-1 text-titles-and-attributes';
+  }
+
+  /**
+   * The options to render inside of the popover list.
+   */
+  get options() {
+    return this.filteredOptions || this.args?.options;
+  }
+
+  /**
+   * The internal currently selected item.
+   */
+  get selected(): string | undefined {
+    let { optionKey, selected } = this.args;
+
+    if (!selected) {
+      return undefined;
+    }
+
+    return (
+      typeof selected === 'object' && optionKey ? selected[optionKey] : selected
+    ) as string;
+  }
+
+  /**
+   * Attempts to scroll the active or newly highlighted item into view for the user.
+   */
+  #scrollActiveOptionIntoView(alignToTop?: boolean) {
+    assert('`this.activeIndex` cannot be `null`', this.activeIndex !== null);
+
+    const optionsElements = document.querySelectorAll(
+      `#${this.popoverId} ${optionComponentSelector}`
+    );
+    const optionElement = optionsElements[this.activeIndex];
+
+    if (!optionElement) {
+      return;
+    }
+
+    optionElement.scrollIntoView(alignToTop);
+  }
+
+  @action
+  closePopover() {
+    this.isPopoverOpen = false;
+  }
+
+  @action
+  noop() {
+    // eslint-disable @typescript-eslint/no-empty-function
+  }
+
+  /**
+   * Action called when a new item is selected. Ultimately calls the provided `@onChange` with the newly selected item.
+   */
+  @action
+  onChange() {
+    assert('`this.activeIndex` cannot be `null`', this.activeIndex !== null);
+
+    assert(
+      '`this.args.options` cannot be  `undefined`',
+      this.args.options !== undefined
+    );
+
+    let { optionKey, onChange } = this.args;
+
+    this.closePopover();
+
+    // This shouldn't be possible, but to satisfy TS
```

Comment:
Nit: If it's not possible to be in that state, then we can use `assert()`, which makes TS happy and does not add runtime code (in prod).

Reply:
Perfect! Pushed as part of commit https://github.com/CrowdStrike/ember-toucan-core/pull/200/commits/4f2d139737aa6ddb6025150574247a191cc1fe03